File: src/exp1_wasm_bench.py

import wasmtime
import time

engine = wasmtime.Engine(wasmtime.Config())
store = wasmtime.Store(engine)
module = wasmtime.Module.from_file(engine, "exp1_wasm_module.wasm")
instance = wasmtime.Instance(store, module, [])

memory = instance.exports(store)["memory"]
decode_seq = instance.exports(store)["decode_seq"]
decode_sel = instance.exports(store)["decode_sel"]

TOTAL = 100000000 // 10
memory.grow(store, max(2000, int(TOTAL * 8 / 65536) + 10))

decode_seq(store, 0, TOTAL * 4, TOTAL) # Warm-up

start = time.perf_counter()
decode_seq(store, 0, TOTAL * 4, TOTAL)
end = time.perf_counter()
seq_mtps = (TOTAL / 1000000.0) / (end - start)

start = time.perf_counter()
decode_sel(store, 0, TOTAL * 4, TOTAL, 10)
end = time.perf_counter()
sel_mtps = (TOTAL / 1000000.0) / (end - start)

print(f"{seq_mtps},{sel_mtps}")

File: src/exp1_wasm_bench.cpp
#include <iostream>
#include <chrono>
#include <wasmtime.hh>

using namespace std;

const size_t TOTAL = 100000000;

int main() {
    wasmtime::Engine engine;
    wasmtime::Store store(engine);
    
    // 加载 Wasm 模块
    wasmtime::Module module = wasmtime::Module::compile(engine, "exp1_wasm_module.wasm").unwrap();
    wasmtime::Instance instance = wasmtime::Instance::create(store, module, {}).unwrap();
    
    // 获取导出函数
    auto decode_seq = instance.get(store, "decode_seq").unwrap().func();
    
    // 分配足够大的线性内存并预热
    auto memory = instance.get(store, "memory").unwrap().memory();
    memory.grow(store, 6000).unwrap(); // 动态申请页
    
    decode_seq.call(store, {0, TOTAL * 4, TOTAL}).unwrap(); // Warm-up
    
    // 纯 C++ 态调用 Wasm 测量真实沙盒税
    auto start = chrono::high_resolution_clock::now();
    decode_seq.call(store, {0, TOTAL * 4, TOTAL}).unwrap();
    auto end = chrono::high_resolution_clock::now();
    
    double secs = chrono::duration<double>(end - start).count();
    double mtps = (TOTAL / 1000000.0) / secs;
    
    cout << "Wasmtime (Pure C++ API) Sequential MT/s: " << mtps << endl;
    return 0;
}